core/server/services/bulk-email/mailgun.js

```diff
@@ -0,0 +1,40 @@
+const {URL} = require('url');
+const mailgun = require('mailgun-js');
+const common = require('../../lib/common');
+const configService = require('../../config');
+const settingsCache = require('../settings/cache');
+
+function createMailgun(config) {
+    const baseUrl = new URL(config.baseUrl);
+
+    return mailgun({
+        apiKey: config.apiKey,
+        domain: config.domain,
+        protocol: baseUrl.protocol,
+        host: baseUrl.host,
+        port: baseUrl.port,
+        endpoint: baseUrl.pathname
+    });
+}
+
+function getInstance() {
+    const bulkEmailConfig = configService.get('bulkEmail');
+    const bulkEmailSetting = settingsCache.get('bulk_email_settings');
+    const hasMailgunConfig = !!(bulkEmailConfig && bulkEmailConfig.mailgun);
+    const hasMailgunSetting = !!(bulkEmailSetting && bulkEmailSetting.apiKey && bulkEmailSetting.baseUrl && bulkEmailSetting.domain);
+    if (!hasMailgunConfig && !hasMailgunSetting) {
```

Comment:
Just ran into a bug where I left the `tag` configuration in the config file like following:
```
    "bulkEmail": {
        "mailgun": {
            "tag": "tag-from-naz-yo"
        }
    }
```
and had dynamic config filled out. With such combo the initialization throws an error and ends up at https://github.com/TryGhost/Ghost/pull/11386/files#diff-494daed86df08277442c46800b6f0e2aR32.

Think the correct way here would be merging two configurations with priority to bulkEmailSetting instead of passing partial config from bulkEmailConfig :thinking: What are your opinions @rishabhgrg  @allouis?

Reply:
🤔 Interesting, the reason it fails atm is because of implicit assumption that if there is `config.mailgun`, then its expected to consist all mailgun values. IMO merging the configurations from 2 places could be more misleading/confusing as user can partially fill values in 2 different places and then trip on it later. 

I think the best solution for this is in case we have `config.mailgun`, then we expect user to have full config there and add an explicit check for it in code, throwing warning "Incorrect/Invalid mailgun config" if its not complete. In case there is no `config.mailgun` we fallback to settings,  which is again expected to have the full set of data which we already check for [here](https://github.com/TryGhost/Ghost/pull/11386/files#diff-494daed86df08277442c46800b6f0e2aR24). This way, if anyone is attempting to fill mailgun data in config,  they are expected to add it all there, and if not, then its picked from settings,  which is more deterministic and less confusing.

Reply:
Having to rely on one or another is a good idea! The downside I see with this approach is that you won't be able to configure tagging in any way when using dynamic settings? Unless we add that in the DB/UI.

Reply:
> that you won't be able to configure tagging in any way when using dynamic settings?
Right, but I feel that's an edge use-case, as config will only be changed by self-hosters to update the tagging, in which case they can always also add rest of the config in it instead of putting it in dynamic settings. This will also lead to hiding the UI completely in admin, so its not confusing where their config is coming from. The case where you go and edit config to add tagging, but not rest of the config is not ideal anyways imo, as the whole thing is part of one mailgun instance setting. :)

Reply:
> I feel that's an edge use-case, as config will only be changed by self-hosters to update the tagging, in which case they can always also add rest of the config in it instead

:thinking: okkie. probably worth throwing some other error in this case. maybe worth throwing a lil more specific error in a case like this one. "missing config key" or something, because "Bulk email service is not configured" was very unclear :smiley:

Reply:
😄 Agreed! Def need better wording for incomplete/incorrect configured config. Does the rest sound sane though?